File: src/main/java/pl/sixpinetrees/tournament/domain/Competition.java
package pl.sixpinetrees.tournament.domain;

import pl.sixpinetrees.tournament.util.Calculator;

import javax.persistence.Embedded;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;

@Entity
public class Competition {

    @Id
    @GeneratedValue
    private Long id;

    private String name;

    private Integer numberOfPlayers;

    private Integer numberOfRounds;

    private Integer numberOfMatchesInFirstRound;

    @Embedded
    private VictoryConditions victoryConditions;

    private Long winningPlayerId;

    public Competition() {
    }

    public Competition(String name, Integer numberOfPlayers, Integer numberOfWinsRequired, Integer numberOfPointsToWin) {

        this.name = name;
        this.victoryConditions = new VictoryConditions(numberOfWinsRequired, numberOfPointsToWin);
        calculateRounds(numberOfPlayers);
    }

    private void calculateRounds(Integer noOfPlayers) {
        numberOfPlayers = noOfPlayers;
        numberOfRounds = Calculator.log2ceil(numberOfPlayers);
        numberOfMatchesInFirstRound = numberOfPlayers - Calculator.pow2N(numberOfRounds - 1);
    }

    public Integer calculateMatchesInRound(Integer round) {
        if (round == 1) {
            return numberOfMatchesInFirstRound;
        }
        return Calculator.pow2N(numberOfRounds - round);
    }

    public void assignWinner(Long playerId) {
        this.winningPlayerId = playerId;
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getNumberOfPlayers() {
        return numberOfPlayers;
    }

    public Integer getNumberOfRounds() {
        return numberOfRounds;
    }

    public Integer getNumberOfMatchesInFirstRound() {
        return numberOfMatchesInFirstRound;
    }

    public Long getWinningPlayerId() {
        return winningPlayerId;
    }

    public VictoryConditions getVictoryConditions() {
        return victoryConditions;
    }
}

File: src/app/competition/competition.ts
import { VictoryConditions } from "./victory-conditions";

export class Competition {
  id: number;
  name: string;
  numberOfPlayers: number;
  numberOfRounds: number;
  numberOfMatchesInFirstRound: number;
  victoryConditions: VictoryConditions;


  calculateMatchesInRound(round: number): number {
    if (round == 1) {
      return this.numberOfMatchesInFirstRound;
    }
    return Math.pow(2, this.numberOfRounds - round);
  }

  calculateMaxMatchesInRound(round: number): number {
    return Math.pow(2, this.numberOfRounds - round);
  }

  isDoubleMatch(round: number, position: number): boolean {
    return (position % 2 == 1 && position < this.calculateMatchesInRound(round));
  }

  isSingleMatch(round: number, position: number): boolean {
    return (position % 2 == 1 && position == this.calculateMatchesInRound(round));
  }

  isWithoutMiddleBar(round: number, position: number): boolean {
    return (round == 1) || (round == 2 && position > (this.numberOfMatchesInFirstRound + 1) / 2);
  }

}
